### DISCUSSION TITLE
How to export data using nbshell?

### DISCUSSION THREAD

**Discussion by hoalex** [netbox-community/netbox#8443] [Q&A] [answered]
Hi there,

I am looking for a way to use the export functionality of the web UI via Nbshell (the standard "All Data" export would be fine; using a custom export template would be even better). Is this even possible, and if yes, does anyone know how to use it?

Thanks!
Alex

**Comment by jeremystretch** [comment, marked as answer]
`nbshell` is just a wrapper around the Python interactive shell, so you can do pretty much anything you can write code for.

You can find the ExportModel class defined in [`extras/models.py`](https://github.com/netbox-community/netbox/blob/69eb6b11d0ee288b41d6b905e890b5995273e092/netbox/extras/models/models.py#L252). This class includes a `render()` method which takes a queryset and returns the rendered output. So, you can do something like this:

```python
sites = Site.objects.all()
et = ExportTemplate.objects.get(name='MyExportTemplate')
et.render(sites)
```

Hope that helps!

**Reply by hoalex** [reply]
Thanks, that works! 👍 
Any way to also run the default "All data" export via `nbshell`? It seems it's not technically defined as an export template...

**Reply by jeremystretch** [reply]
The default export is handled by the table class. You can find the code for it [here](https://github.com/netbox-community/netbox/blob/69eb6b11d0ee288b41d6b905e890b5995273e092/netbox/netbox/views/generic.py#L182).

**Reply by hoalex** [reply]
Thank you for the pointer, I will see if I can get that to work. For the moment, using the export templates works fine for me.

**Reply by jonasnieberle** [reply]
@jeremystretch Is it possible that this approach doesn't work with the netbox version v4.4.6?

The export template work in the netbox UI, but not with the nbshell.

```
>>> manufacturers = Manufacturer.objects.all()
>>> test = ExportTemplate.objects.get(name="export_manufacturers")
>>> test.render(manufacturers)
Traceback (most recent call last):
  File "/python3.11/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
  File "/netbox-4.4.6/netbox/extras/models/mixins.py", line 142, in render
    context = self.get_context(context=context, queryset=queryset)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/netbox-4.4.6/netbox/extras/models/models.py", line 454, in get_context
    _context.update(context)
TypeError: cannot convert dictionary update sequence element #0 to a sequence
```

**Reply by jonasnieberle** [reply]
OK - I got it:

```
>>> test.render(queryset=manufacturers)
```
